### PR TITLE
Add comparison support

### PR DESCRIPTION
**Pull request by Sollace**
This adds support for the "setDefaultComparison(thing, comparison)" method.

It's a little different from how EMI does it, instead of `context.setDefaultComparison(item, comparison)` you do `context.getItemComparisons().register(item, comparison)` because I wanted to allow for overrides that let you register multiple at once to enable code like:

```
        context.getItemComparisons().register(TlaStackComparison.compareComponents(),
                PSItems.WOODEN_MUG, PSItems.STONE_CUP, PSItems.GLASS_CHALICE,
                PSItems.SHOT_GLASS, PSItems.BOTTLE, PSItems.MOLOTOV_COCKTAIL, PSItems.SYRINGE,
                PSItems.FILLED_GLASS_BOTTLE, PSItems.FILLED_BUCKET, PSItems.FILLED_BOWL
        );
        context.getFluidComparisons().register(TlaStackComparison.compareComponents(), SimpleFluid.REGISTRY);
```

If you think it's more important to be binary compatible with EMI's methods I'm open to adding back the `setDefaultComparison` methods I had in an earlier revision.

Submitting as a draft whilst I test it in an environment with a mod that would use this.

### CODE REVIEW

File: src/client/java/io/github/mattidragon/tlaapi/api/plugin/Comparisons.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+     * @param comparison The comparison and hash function
+     */
+    void register(T key, TlaStackComparison comparison);
+
+    /**
+     * Adds a default comparison method for a stack for its key.
+     * @param keys The keys for the object to compare.
+     * @param comparison The comparison and hash function
+     */
+    default void register(TlaStackComparison comparison, @SuppressWarnings("unchecked") T...keys) {
```
Comment (MattiDragon) [inline src/client/java/io/github/mattidragon/tlaapi/api/plugin/Comparisons.java]: Formatting issue: There should be a space after the varargs dots

File: src/client/java/io/github/mattidragon/tlaapi/api/plugin/Comparisons.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            register(key, comparison);
+        }
+    }
+
+    /**
+     * Adds a default comparison method for a stack for its key.
+     * @param keys The keys for the object to compare.
+     * @param comparison The comparison and hash function
+     */
+    default void register(TlaStackComparison comparison, Stream<T> keys) {
```
Comment (MattiDragon) [inline src/client/java/io/github/mattidragon/tlaapi/api/plugin/Comparisons.java]: Is this actually ever useful? I'd think it's niche enough that you can just call `toList` and pass it to the overload above.

File: implSrc/jei/java/io/github/mattidragon/tlaapi/impl/jei/TlaApiJeiPlugin.java [hunk 3]
```diff
@@ -242,13 +275,24 @@ public <T extends Screen> void addScreenSizeProvider(Class<T> clazz, Function<T,
         sizeProviders.add(new ScreenSizeProvider<>(clazz, provider));
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public <T extends Screen> void addExclusionZoneProvider(Class<T> clazz, Function<T, ? extends Iterable<TlaBounds>> provider) {
         if (!HandledScreen.class.isAssignableFrom(clazz)) return; // JEI only supports exclusion zones on handled screens
         //noinspection unchecked
```
Comment (MattiDragon) [inline implSrc/jei/java/io/github/mattidragon/tlaapi/impl/jei/TlaApiJeiPlugin.java]: As you added the supress to the method, please remove the comment supression

File: src/client/java/io/github/mattidragon/tlaapi/api/recipe/TlaStackComparison.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package io.github.mattidragon.tlaapi.api.recipe;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * A combined equality and hashing function for TlaStacks.
+ * <p>
```
Comment (MattiDragon) [inline src/client/java/io/github/mattidragon/tlaapi/api/recipe/TlaStackComparison.java]: Why did you add a paragraph split here? I don't think it's necessary

File: src/client/java/io/github/mattidragon/tlaapi/api/recipe/TlaStackComparison.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    public static TlaStackComparison of(Predicate comparator, HashFunction hashFunction) {
+        return new TlaStackComparison(comparator, hashFunction);
+    }
+
+    public static <T> TlaStackComparison compareData(Function<TlaStack, T> function) {
+        return of((a, b) -> Objects.equals(function.apply(a), function.apply(b)), stack -> Objects.hashCode(function.apply(stack)));
+    }
+
+    public interface Predicate {
```
Comment (MattiDragon) [inline src/client/java/io/github/mattidragon/tlaapi/api/recipe/TlaStackComparison.java]: I'd make this extend `BiPredicate`. Maybe also rename `EqualityPredicate` to differenciate from the jdk interface

File: gradle.properties [hunk 6]
```diff
@@ -5,7 +5,7 @@ minecraft_version=1.21
 yarn_mappings=1.21+build.2
 loader_version=0.15.11
 
-mod_version=1.2.0
+mod_version=1.2.1
```
Comment (MattiDragon) [inline gradle.properties:8]: If you're bumping this you should also create a new changelog file and mark your changes there

**Review by MattiDragon — CHANGES_REQUESTED** [review]
Some minor issues, but overall design looks good to me

File: gradle.properties (hunk 6 above)
Comment (Sollace) [reply gradle.properties:8]: Alright! I bumped it so I could make a build.

I'll add that to the PR after the other stuff.

File: src/client/java/io/github/mattidragon/tlaapi/api/plugin/Comparisons.java (hunk 2 above)
Comment (Sollace) [reply src/client/java/io/github/mattidragon/tlaapi/api/plugin/Comparisons.java]: Yeah fair. I thought I would include it because Streams don't implement the Iterable<T> interface (which I find weird). I'll get rid of it

File: implSrc/jei/java/io/github/mattidragon/tlaapi/impl/jei/TlaApiJeiPlugin.java (hunk 3 above)
Comment (Sollace) [reply implSrc/jei/java/io/github/mattidragon/tlaapi/impl/jei/TlaApiJeiPlugin.java]: The comment suppression didn't do anything so I added it. Will get rid of it

File: src/client/java/io/github/mattidragon/tlaapi/api/recipe/TlaStackComparison.java (hunk 5 above)
Comment (Sollace) [reply src/client/java/io/github/mattidragon/tlaapi/api/recipe/TlaStackComparison.java]: These are the names EMI used for their one. I didn't see any reason why I should change them

**Comment by Sollace** [comment]
Sorry about that @MattiDragon . I've made all the requested changes if you want to look at this again. :)

**Review by MattiDragon — APPROVED** [review]
